### ISSUE TITLE
CHANGE: `inheritFrom` to use template name and not label

### ISSUE THREAD

**Issue by chb0github** [jenkinsci/amazon-ecs-plugin#174]
Having had some first hand experience with this, I think the `inheritFrom` mechanism should only be based on the template name. There are a couple of reasons:

1. Sheer intuition. I inherit from my parents, who have names, not labels. In addition, the word `inherit` in the Java/programming world is very clear. In Java I can only inherit from 1 parent. People have a clear understanding of the word inheritance and by actually actually inheriting from `label` it's utterly baffling. What if I have 2 templates with the same label and utterly different properties? It's really a function of iteration order inside the code which you actually get.
2. The symantecs of labeling agents and their use is already clearly understood and uses an entirely different mechanism:

```groovy
agent{
   label 'my-agent-label'
}
```
And there is no indication at all this is related to ECS (as is intended).

I propose: 

1. `inheritFrom` only uses **only** template name. it make the match logic way simpler. It's a 1-line lookup
2. remove any usage of `label` in the ecs script definition. If you want to use an agent by label there is already that mechanism and it's well established. The labeling can still exist in the template cloud definition, but it's simply not something the script can call upon
3. Ideally, in the ecs agent code, if you just say 
```
agent{
  ecs 'some-template'
}
```
or
```groovy
agent{
   ecs 
}
```
then you will get `cloud-default`-`template-default`

Forgive me if I offend any original authors, but this plugin is extremely confusing to use.

**Comment by webratz** [comment]
@bruecktech was originally adding the support. maybe he can share some things :)
in general: this would be a breaking change, so imho we should have first a version that adds the new behavior and introduces deprecation messages for the old one. and after a while then remove it

**Comment by chb0github** [comment]
Well, I wrote code with this last push to support a few of these features. To be honest, the intent is to break/change it. It has all sorts of ambiguous execution paths. 

The current functionality is extremely fragile. 
it basically looks for the first cloud that contains a label matching it's `inheritFrom`. If the order changes: Broke. If implementation of the cloud list changes, OR you even sort it: Broke (and this one is nasty because this is off in the bowels of Jenkins, they would never know they broke something downstream). The current implementation is way too susceptible to many external forces. 

Consider these scenarios and what behavior you're reasonably expect?

## Scenario 1 
- several clouds, different names. Each many labels. 3 of them have an identical label amongst many (making this situation hard to spot). However, each cloud is totally configured different. In fact, each one uses a different key/iam and thus _bills_ to that key. So, I am a dev, I write some code that say: `inheritFrom 'duplicate-label'` My admin comes in, adds an extra label to 1 of the clouds (whichever he needs). What would you expect? 
1. Nothing changes?
2. My jobs now start to build somewhere totally different and fail or worse, cause damage or cost money

## scenario 2
Somewhere off in the bowels of the Jenkins code, maybe in another plugin, `Jenkins.instance.clouds` gets reordered. By accident, by intent. The list isn't immutable or even controlled. Infact, it's a `public final Jenkins.CloudList cloud` So, there isn't even a way to control access. What would you expect?
1. Nothing happens
2. Jobs from an unrelated plugin/code now start to build against a different cloud and then fail/damage/delete/cost ?


As for adding the new along side of old: That would actually add to the confusion, I think. You can't BOTH inheritFrom a label and `inheritFrom` a name - If you have 1 or the other, maybe that makes some sense. But you can't have both. And that's a case to account for. You might be able to _prefer_ one first, then the other, but this too would be a breaking change potentially. And still confusing. 

Which also brings up another reason it doesn't make sense to use labels: I don't know how this code will behave if you have both a label and `inheritFrom`

```groovy
ecs{
  label 'foo'
  inheritFrom 'baz'
}
```
What would this code even mean at a glance? 

I understand where you're coming from. I am a seasoned veteran and I would normally agree with you. But this plugin in unintuitive and fragile. We'd be doing the community a favor by just showing 2 clear examples of how to use it and deleting a field called `label` from the `ecs{}` block:

```groovy
ecs{
  inheritFrom 'some-template-name'
}
```
or
```groovy
agent{
  label 'foo'
}
```

**Comment by webratz** [comment]
i see. I'd propose we do the following: We add some code to the current state, that notifies people there will be an upcoming breaking change (if they use inheritfrom), so we are not breaking what some might have gotten working without telling them.
Next step would be to make the actual change (without keeping compatibility, just make the new way work)

Let's maybe also make the naming more explicit, eg

```
ecs{
  inheritFromTemplate 'some-template-name'
}
```

**Comment by chb0github** [comment]
I was considering something along the same line; we could even put a warning in the UI as well as the logs. 

Are for the naming: I can definitely get on board with something more intuitive:

```
ecs{
  parentTemplate 'some-template'
}
```
```
ecs{
  fromTemplate 'some-template'
}
```
The only reason I have these suggestion is because I feel yours is a bit wordy, but honestly, I won't put up a fight.

**Comment by webratz** [comment]
as long it includes template i'm fine i think. just to make it really explicit on what is used there

**Comment by wuillaum** [comment]
I disagree with removing the ability to specify a label in dynamic (pipeline) templates. I use it to create dynamic templates for 1-off tasks on thousands of job-runs per day. It's especially necessary if using inheritFrom, because without overriding your parent's label there isn't a guarantee that your pipeline will have the newly created dynamic task provisioned, and instead the parent could be provisioned unintentionally. If a randomized label was created by default when "merging" the templates then I don't see a problem, but users should still have the ability to override with their own naming scheme by using `label`.

In cases where I use a taskDefinitionOverride it doesn't make sense to use inheritFrom, and I still want the ability to control my own labels.

I do agree that inheritFrom should use names by default, and that it makes sense to recommend names over labels in the documentation. I think it would be best to keep both features for now.

**Comment by wuillaum** [comment]
Going through the code I think I'm mistaken; disregard my previous comment until I've updated this one... I'm a lot more confused now than I was before (I believe I confusing ecsTaskTemplate() with ecs{}, but I'll be back to confirm).

**Comment by wuillaum** [comment]
I've updated the above comment with my thoughts.

**Comment by chb0github** [comment]
@wuillaum 

the idea is simple: If you want to use the cloud agent by label, the mechanism for doing so already exists:
```groovy
agent{
  label 'mylabel'
}
```
And, this will pick and agent (cloud or otherwise) with that label and run your job. When you use a label, you're not being fussy about where it runs.

However, in the case of cloud/templates, you're being extremely fussy - The task role, the execution role, security groups, etc. By selecting a specific template you are specifically selecting a bundle of configuration that can directly impact costs/permissions and you are doing so with full awareness: I propose this can be done with syntax like:

```groovy
agent{
  ecs{
    template 'customer-cloud'
  }
}
```

This very clearly would charge my customer for this build. And, since template names are unique, there is no possibility for it to just run any-old-place. 

For the time being, yes, agreed, both would run in parallel with a deprecation plan. However, this discussion is good as it exposes the lack of deep semantic thought around the original design/implementation.
